nexus/src/app/background/physical_disk_adoption.rs

```diff
@@ -0,0 +1,110 @@
+//! Oxide racks. When this happens, if they had not previously
+//! been part of a cluster, they need to be explicitly added
+//! to become usable.
+//!
+//! In the future, this may become more explicitly operator-controlled.
+
+use super::common::BackgroundTask;
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use nexus_db_model::PhysicalDisk;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::DataStore;
+use nexus_types::inventory::Collection;
+use serde_json::json;
+use std::sync::Arc;
+use tokio::sync::watch;
+use uuid::Uuid;
+
+pub struct PhysicalDiskAdoption {
+    datastore: Arc<DataStore>,
+    rx_inventory_collection: watch::Receiver<Option<Collection>>,
+}
+
+impl PhysicalDiskAdoption {
+    pub fn new(
+        datastore: Arc<DataStore>,
+        rx_inventory_collection: watch::Receiver<Option<Collection>>,
+    ) -> Self {
+        PhysicalDiskAdoption { datastore, rx_inventory_collection }
+    }
+}
+
+impl BackgroundTask for PhysicalDiskAdoption {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        async {
+            let mut disks_added = 0;
+            let log = &opctx.log;
+            warn!(&log, "physical disk adoption task started");
+
+            let collection = self.rx_inventory_collection.borrow().clone();
+            let Some(collection) = collection else {
+                warn!(
+                    &opctx.log,
+                    "Physical Disk Adoption: skipped";
+                    "reason" => "no inventory"
+                );
+                return json!({ "error": "no inventory" });
+            };
+
+            for (id, _) in collection.sled_agents {
+                // TODO: Make sure "not found" doesn't stop execution
+                let result = self.datastore.physical_disk_uninitialized_list(
+                    opctx,
+                    collection.id,
+                    id,
+                ).await;
+
+                let uninitialized = match result {
+                    Ok(uninitialized) => uninitialized,
+                    Err(err) => {
+                        warn!(
+                            &opctx.log,
+                            "Physical Disk Adoption: failed to query for insertable disks";
+                            "err" => %err,
+                        );
+                        return json!({ "error": format!("failed to query database: {:#}", err) });
+                    },
+                };
+
+                for inv_disk in uninitialized {
+                    let disk = PhysicalDisk::new(
+                        Uuid::new_v4(),
+                        inv_disk.vendor,
+                        inv_disk.serial,
+                        inv_disk.model,
+                        inv_disk.variant,
+                        inv_disk.sled_id,
+                    );
+
+                    let result = self.datastore.physical_disk_insert(opctx, disk).await;
```

Comment:
Just making sure I understand the flow; is this right?

1. A new sled is added
2. The sled-agent on the new sled creates zpools on all its physical disks
3. An inventory collection is taken when includes the new sled, which includes zpools+physical disks from step 2
4. This background task executes, which adopts any zpool+physicaldisk pairs, turning them into control plane `PhysicalDisks` via this call
5. The next time the planner runs, its `PlanningInput` includes the new sled with its zpool+physicaldisk pairs (it already would have gotten the new zpools; getting the physical disk alongside it is new)
6. The planner uses the new disk input to build `blueprint_disks`
7. The executor sends the list of zpool+disk pairs for each sled from `blueprint_disks` to every sled-agent (an idempotent operation)

This gets at the question of autoadoption vs operator-driven adoption - it feels a little backwards to me that the disks are adopted _before_ reconfigurator gets involved, and then execution is only responsible for telling sled-agents about it. I think the feeling is because for external networking resources, reconfigurator chooses whether to assign resources during planning, and then execution is responsible for both (a) record that decision in the database and then (b) tell sled-agent about it, whereas for disks (a) is happening ahead of time. Maybe that's fine or even better in this case? I'm not sure, but just wanted to jot down the feeling.

Reply:
I think it makes sense to auto adopt in this case. We need a mechanism to create the physical control plane disks so that reconfigurator can see them. This is somewhat of a stop-gap because we don't have APIs to list raw disks and allow the operator to adopt them. However, as an intermediate step, we could choose to have the physical control plane disks not put in service and then adopted by reconfigurator. I wouldn't worry about that for this PR though.